core/src/test/scala/org/apache/spark/deploy/DecommissionWorkerSuite.scala

```diff
@@ -0,0 +1,424 @@
+          // are actually on the same host)
+          throw new FetchFailedException(BlockManagerId(executorToDecom,
+            workerToDecom.host, workerToDecom.port), 0, 0, -1, 0, "Forcing fetch failure")
+        }
+        val sumVal: List[Int] = List(iter.sum)
+        sumVal.iterator
+      }, preservesPartitioning = true)
+      .sum()
+    assert(jobResult === 2)
+    // 6 tasks: 2 from first stage, 2 rerun again from first stage, 2nd stage attempt 1 and 2.
+    val tasksSeen = listener.getTasksFinished()
+    assert(tasksSeen.size === 6, s"Expected 6 tasks but got $tasksSeen")
+  }
+
+  private abstract class RootStageAwareListener extends SparkListener {
+    protected var rootStageId: Option[Int] = None
+    var tasksFinished = new ConcurrentLinkedQueue[String]()
+    var jobDone = new AtomicBoolean(false)
+
+    // We cannot use regular `assert` reliably inside the listener onXXX methods because they
+    // are called in a separate thread, and exceptions within that thread are not propagated
+    // upto the main test thread to trigger a test failure. Instead we queue up
+    // assertion failures and then assert we haven't seen any at the end.
+    // assertionFailures is a queue of (assertionMessage, callingMethodStack)
+    var assertionFailures = new ConcurrentLinkedQueue[(String, String)]()
+
+    protected def isRootStageId(stageId: Int): Boolean =
+      (rootStageId.isDefined && rootStageId.get == stageId)
+
+    override def onStageSubmitted(stageSubmitted: SparkListenerStageSubmitted): Unit = {
+      if (stageSubmitted.stageInfo.parentIds.isEmpty && rootStageId.isEmpty) {
+        rootStageId = Some(stageSubmitted.stageInfo.stageId)
+      }
+    }
+
+    def delayedAssert(condition: Boolean, message: => String): Unit = if (!condition) {
+      assertionFailures.add((message, Thread.currentThread().getStackTrace()(2).toString))
+    }
+
+    override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
+      jobDone.set(true)
+    }
+
+    protected def handleRootTaskEnd(end: SparkListenerTaskEnd) = {}
+
+    protected def handleRootTaskStart(start: SparkListenerTaskStart) = {}
+
+    override def onTaskStart(taskStart: SparkListenerTaskStart): Unit = {
+      if (isRootStageId(taskStart.stageId)) {
+        handleRootTaskStart(taskStart)
+      }
+    }
+
+    override def onTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = {
+      val taskSignature = s"${taskEnd.stageId}:${taskEnd.stageAttemptId}:" +
+        s"${taskEnd.taskInfo.index}:${taskEnd.taskInfo.attemptNumber}"
+      logInfo(s"Task End $taskSignature")
+      tasksFinished.add(taskSignature)
+      if (isRootStageId(taskEnd.stageId)) {
+        handleRootTaskEnd(taskEnd)
+      }
+    }
+
+    def getTasksFinished(waitTime: Span = 10.seconds): Seq[String] = eventually(timeout(waitTime)) {
+      val done = jobDone.get()
+      assert(done)
+      assert(assertionFailures.isEmpty, s"Delayed asserts: $assertionFailures")
```

Comment:
The `getTasksFinished` checks the delayed asserts as a side effect. I think either a new name (which suggest the asserts are checked) or a separate method would be better.